CMakeLists.txt

```diff
@@ -189,6 +174,45 @@ macro(check_compiler file_to_test flag_to_set)
   endif()
 endmacro()
 
+macro(check_run file_to_test flag_to_set compile_flags)
+  message(STATUS "Run check_run file_to_test '${file_to_test}', "
+          "flag_to_set '${flag_to_set}', "
+          "and compile_flags '${compile_flags}'.")
+  try_run(
+          ${flag_to_set}
+          COMPILE_RESULT
+          "${CMAKE_CURRENT_BINARY_DIR}"
+          "${CMAKE_CURRENT_LIST_DIR}/tests/compiler_features_tests/${file_to_test}"
+          COMPILE_DEFINITIONS "${compile_flags}"
+          OUTPUT_VARIABLE COMPILE_AND_RUN_OUTPUT)
+  if (NOT COMPILE_RESULT)
+    message(WARNING "COMPILE_AND_RUN_OUTPUT ${COMPILE_AND_RUN_OUTPUT}")
+  endif()
+endmacro()
+
+# Detect if memcmp is wrongly stripped like strcmp.
+# If exists, let CMake generate a warning.
+# memcmp bug link https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95189.
+if (GCC)
+  # CMake try_run requires these variables must be preset.
+  # https://cmake.org/cmake/help/latest/command/try_run.html
+  set(MEMCMP_INVALID_STRIPPED "")
+  set(MEMCMP_INVALID_STRIPPED__TRYRUN_OUTPUT "")
+  if (CMAKE_BUILD_TYPE_LOWER MATCHES "release")
+    # CMAKE_C_FLAGS_RELEASE enables `-O3`.
+    check_run(memcmp_invalid_stripped_check.c MEMCMP_INVALID_STRIPPED "${CMAKE_C_FLAGS_RELEASE}")
+  elseif(CMAKE_BUILD_TYPE_LOWER MATCHES "relwithdebinfo")
+    # CMAKE_C_FLAGS_RELEASE enables `-O2`.
+    check_run(memcmp_invalid_stripped_check.c MEMCMP_INVALID_STRIPPED "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
```

Comment:
Do `CMAKE_C_FLAGS_RELEASE` and `CMAKE_C_FLAGS_RELWITHDEBINFO` contain flags that are different from the standard `CMAKE_C_FLAGS`?

Reply:
yes, when `CMAKE_BUILD_TYPE` is set to `Release` or `RelWithDebInfo`, the optimized compiler flag (mentioned in code) will be used to compile code. This can be verified by enabling CMake verbose flag. This is another reason why the compiler issue does not happen when building with DEBUG mode.